3단계 - 질문 삭제하기 리팩터링

ini8262 opened the pull request.

ini8262 committed:
refac : 엔티티 클래스 리팩토링

- Owner / Deleteable 인터페이스 추가
- Getter 메소드 변경

ini8262 committed:
refac : 엔티티간의 관계 및 삭제처리 리팩토링

- Question과 Answer간의 편의함수 조정
- Deleteable.class 리팩토링
- Answers와 DeleteHistories 일급객체 추가

ini8262 committed:
test : 엔티티 리팩토링으로 테스트케이스들도 리팩토링

ssosso commented on src/main/java/qna/domain/entity/DeleteHistory.java:
```diff
@@ -37,19 +32,32 @@
+        }
+
+        protected DeleteHistory getDeleteHistory(Long contentId, User user) {
+            return new DeleteHistory(this, contentId, user, LocalDateTime.now());
```
👍

ini8262 replied:
😃

ssosso commented on src/main/java/qna/domain/entity/common/Deleteable.java:
```diff
@@ -1,5 +1,44 @@
+public abstract class Deleteable<T, R> extends TraceDate {
+
+    @Column(nullable = false)
+    private boolean deleted;
```
👍

ini8262 replied:
😃

ssosso commented on src/main/java/qna/domain/entity/Question.java:
```diff
@@ -39,15 +36,15 @@
+    private Answers answers = new Answers();
+
+    @Transient
+    private final DeleteHistories deleteHistories = new DeleteHistories();
```
상태값으로 가지고 있어야 할 정보일까요~?

ini8262 replied:
해당부분 코드작성할때, 고민했던 사항이였습니다.  
상태값으로 가지고 있게 구현한 이유는 삭제를 취소하는 경우 또는 삭제 지연 처리등으로 발전할수도 있다고 생각하였습니다. 
해당 부분은 수정하여 커밋하였습니다.

ssosso replied:
그런 상황을 고려하셨었군요~ 좋은 아이디어네요:)
다만 이전에 구현된대로 두면 deleted 메서드 호출 시 answers에서 exception이 발생하면 deleteHistories에 question에 대한 history가 저장된 상태로 남을거 같아서요~ 고려했던 요구사항이 생긴다면 그때 고려하여 설계해도 좋을거 같아요:)

ssosso commented on src/test/java/qna/domain/entity/AnswerTest.java:
```diff
@@ -31,9 +35,9 @@ void isOwner() {
+        ANSWER_OF_SANJIGI.deleted(UserTest.USER_SANJIGI);
 
         assertAll(
             () -> assertThat(ANSWER_OF_SANJIGI.isDeleted()).isTrue()
```
주인이 삭제하지 않는 경우에 대한 테스트도 있으면 좋을거 같아요~

ini8262 replied:
꼼꼼한 테스트 명심하겠습니다!

ssosso reviewed (changes requested):
태훈님~ 피드백 드리오니 확인 부탁드립니다:)

ini8262 committed:
test : 잘못된 given 데이터 수정 및 테스트케이스 추가

- 질문 삭제 가능 리스트 조회
-  다른 작성자가 작성한 답변을 삭제하는 경우 예외가 발생
- 질문 삭제 가능 리스트중에서 질문 하나를 정상 삭제후 질문 리스트 확인

ini8262 committed:
refac : Question 엔티티에 불필요한 상태값 제거

ssosso reviewed (approved):
태훈님~ 수정해주시고 댓글 달아주신 내용 모두 확인하였습니다:)
고려하신 상황이 있으신 것을 이해하였고 댓글도 알아두었으니 확인 부탁드립니다~
이번 미션은 이만 마무리해도 될 거 같아요, 그동안 고생 많으셨습니다:) 다음 미션 진행 부탁드립니다~
